### PR TITLE
Prototype of a metrics mechanism that intercepts finished spans and r…

### PR DESCRIPTION
**Pull request by objectiser**
…eports their details to a metrics reporter - with an initial prometheus impl

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay @jpkrohling If you could review please.

Notes:
* Currently uses a Tracer wrapper to capture the information, although subject to the outcome of the TracerObserver discussion, it would change to use that. This is an internal impl detail so shouldn't affect the usage of this project.
* The metric labels are configurable - so currently provides just three standard labels (operation, span.kind and error) - but for most tracers the 'service' would also be included as well as potentially others. The readme shows how a custom 'error' label provider could be provided to provide more detailed values.
* Metric labels can also be created based on baggage values - for example, metrics could be partitioned based on business transaction, if the transaction name was carried in the baggage

File: opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * This method transforms the supplied label to ensure it conforms to the
+     * regex "[a-zA-Z_:][a-zA-Z0-9_:]*".
+     *
+     * @param label
+     * @return
+     */
+    @Override
+    protected String convertLabel(String label) {
+        StringBuilder builder = new StringBuilder(label);
```
Comment (jpkrohling) [inline opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java]: Why do you need this conversion?

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            if (!SPAN_LABEL_NAMES.contains(label.name())) {
+                labelNames[pos] = convertLabel(label.name());
+                metricLabels[pos] = label;
+                pos++;
+            } else {
+                overrideTags.put(label.name(), label);
+            }
+        }
+
+        for (int i=0; i < SPAN_LABEL_NAMES.size(); i++){
```
Comment (jpkrohling) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: Without a context, I'm getting a hard time understanding what you are doing here, and why.  Could you add some comments?

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            } else if (OPERATION_LABEL.equals(name)) {
+                metricLabels[pos+i] = new OperationMetricLabel();
+            } else {
+                metricLabels[pos+i] = new TagMetricLabel(name, SPAN_LABEL_DEFAULTS.get(i));
+            }
+        }
+    }
+
+    int numOfTags(List<MetricLabel> tags) {
+        int num = SPAN_LABEL_NAMES.size();
```
Comment (jpkrohling) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: Wouldn't it be nice if we could put Java 7 to rest ? A stream would have been beautiful here :)

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsTracer.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    @Override
+    public <C> void inject(SpanContext context, Format<C> format, C carrier) {
+        tracer.inject(context, format, carrier);
+    }
+
+    void spanStarted(BaseSpan<?> span, String operationName, long startNanoTime, Map<String,Object> tags) {
+        if (reporter != null) {
+            synchronized (metricData) {
```
Comment (jpkrohling) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsTracer.java]: Java 8 could also make this prettier:

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicReference.html#updateAndGet-java.util.function.UnaryOperator-

File: opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java (hunk 1 above)
Comment (objectiser) [inline opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java]: Prometheus has specific restrictions on the format of the label - as shown in the method comment.

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java (hunk 3 above)
Comment (objectiser) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: Agree, but currently there seems to be a Java 6 constraint - until that is lifted we need to make sure any 'added value' components also conform. Unless we potentially restrict where these components can be used?

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsTracer.java (hunk 4 above)
Comment (objectiser) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsTracer.java]: Comment as above.

File: opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java (hunk 1 above)
Comment (jpkrohling) [inline opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java]: Perhaps just expand the comment then, saying that this is a Prometheus requirement?

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java (hunk 2 above)
Comment (objectiser) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: The missing context is due to the prometheus impl - prom requires a fixed list of labels used for a particular metric type - so when the reporter is initialized it combines the metric/labels provided on instantiation, with the standard ones, to provide a definitive list.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,118 @@
...
+by the span's baggage, and if not found uses a default value.
+
+
+### Default Labels
+
+By default, the metrics are reported using the following labels:
+
+* operation - the operation associated with the span
+
+* spen.kind - the _span.kind_ tag associated with the span, by default if not specified, then the metrics
```
Comment (pavolloffay) [inline README.md]: span

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java (hunk 2 above)
Comment (objectiser) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: Let me know if that clears things up - once clear I will add a relevant comment to the code.

File: opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                .labelNames(labelNames).register(registry);
+    }
+
+    @Override
+    public void reportSpan(BaseSpan<?> span, String operation, Map<String, Object> tags, long duration) {
+        String[] labelValues = getLabelValues(span, operation, tags, duration);
+        if (labelValues != null) {
+            this.count.labels(labelValues).inc();
+            // Convert microseconds to seconds
+            this.duration.labels(labelValues).observe(duration / (double)1000000);
```
Comment (pavolloffay) [inline opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java]: nit use TimeUnit

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+import io.opentracing.contrib.metrics.label.TagMetricLabel;
+
+public abstract class AbstractMetricsReporter implements MetricsReporter {
+
+    private static final String OPERATION_LABEL = "operation";
+
+    protected final String[] labelNames;
+    protected final MetricLabel[] metricLabels;
+
+    protected final static List<String> SPAN_LABEL_NAMES = Arrays.asList(OPERATION_LABEL, "span.kind", "error");
```
Comment (pavolloffay) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: List of tuples?

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java (hunk 2 above)
Comment (jpkrohling) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: Makes sense now!

Comment (pavolloffay) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/AbstractMetricsReporter.java]: perhaps extract to a separate method

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/MetricLabel.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+    /**
+     * This method returns a metric tag value.
+     *
+     * @param span The span
+     * @param operation The operation
+     * @param tags The tags
+     * @return The value, if null will suppress the metrics for the span being reported
+     */
+    Object value(BaseSpan<?> span, String operation, Map<String,Object> tags);
```
Comment (pavolloffay) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/MetricLabel.java]: `tags` - shouldn't it be `<String, String>`?

File: opentracing-metrics/src/test/java/io/opentracing/contrib/metrics/AbstractMetricsReporterTest.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        Map<String,Object> spanTags = new HashMap<String,Object>();
+        spanTags.put(Tags.ERROR.getKey(), true);
+        spanTags.put(Tags.HTTP_STATUS.getKey(), 401);
+        spanTags.put(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT);
+
+        String[] labelValues = reporter.getLabelValues(span, "testop", spanTags, 100000L);
+
+        assertEquals(labelValues[0], "TestService");
+        // Adjust index of other tags due to the 'service' tag being added when reporter instantiated
+        assertEquals(labelValues[AbstractMetricsReporter.SPAN_LABEL_NAMES.indexOf(Tags.ERROR.getKey())+1], "4xx");
```
Comment (pavolloffay) [inline opentracing-metrics/src/test/java/io/opentracing/contrib/metrics/AbstractMetricsReporterTest.java]: this look bit magic

File: opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java (hunk 6 above)
Comment (objectiser) [inline opentracing-metrics-prometheus/src/main/java/io/opentracing/contrib/metrics/prometheus/PrometheusMetricsReporter.java]: It needs to be a `double` - TimeUnit returns a `long`.

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/MetricLabel.java (hunk 8 above)
Comment (objectiser) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/MetricLabel.java]: Its `<String,Object>` as the span tags could have non-string values. How these are converted to the representation required by the metrics reporter is an impl specific issue - i.e. prometheus requires string, but other may not.

**Comment by objectiser** [comment]
@pavolloffay @jpkrohling Thanks for the comments so far - I've reworked the way the standard metric labels are managed based on Pavol's tuple comment - it simplifies the initialisation. 

Let me know if updates are ok.

**Comment by objectiser** [comment]
@jpkrohling @pavolloffay Are the latest changes good enough to move forward? i.e. request a ot-contrib repo?

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsBaseSpan.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    @Override
+    public T log(long timestampMicroseconds, String event) {
+        span.log(timestampMicroseconds, event);
+        return (T)this;
+    }
+
+    @Override
+    public T setBaggageItem(String arg0, String arg1) {
+        // TODO Auto-generated method stub
+        return null;
```
Comment (pavolloffay) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsBaseSpan.java]: shouldn't this return something?

File: opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsBaseSpan.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    public T setTag(String key, Number value) {
+        span.setTag(key, value);
+        tracer.spanUpdateTag(span.context(), key, value);
+        return (T)this;
+    }
+
+    @Override
+    public T log(String eventName, Object payload) {
+        // TODO Auto-generated method stub
+        return null;
```
Comment (pavolloffay) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsBaseSpan.java]: shouldn't this return something?

Comment (objectiser) [inline opentracing-metrics/src/main/java/io/opentracing/contrib/metrics/tracer/MetricsBaseSpan.java]: Both done - good catch :)

**Pull request merged**